processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerColumnar.java

```diff
@@ -239,6 +239,7 @@ public void addDataToStore(CarbonRow row) throws CarbonDataWriterException {
    * @return false if any varchar column page cannot add one more value(2MB)
    */
   private boolean isVarcharColumnFull(CarbonRow row) {
+    //TODO: test and remove this as now  UnsafeSortDataRows can exceed 2MB
```

Comment:
Original implementation use `2MB` to ensure next varchar column value can be filled safely, because size of value of single column won't exceed size of a row.
If UnsafeSortDataRows can exceed 2MB(growing dynamically), then we cannot check whether we have enough space for next value because we are not sure how many space next value will take. So the column page size check should be run before adding row to `dataRows`

Reply:
I am not sure how we come to the conclusion of 2MB. There is no guarantee that we always sort the data to use UnsafeSortDataRows always. How about nosort case? And how about if user wants to add 100MB varchar how to support it. 
And also this is not just limited to varchar, we should consider for complex and string columns as well here.
@ajantha-bhat Please remove that todo, But we need to refactor the code to ensure to keep the page size within the snappy max compressed length for complex and string datatypes as well.

Reply:
@xuchuanyin @kevinjmh  @ravipesala @kumarvishal09 : As per discussion let us handle this with configurable page size [from 1 MB to 2GB(snappy max)] and split the complex child pages here only and add validation for each column based on row,

This will be analyzed more and I will open a discussion in community and separate PR will be raised for this.